src/services/db/GitFileSystemService.ts

```diff
@@ -1236,6 +1241,131 @@ export default class GitFileSystemService {
       })
   }
 
+  // Delete multiple files
+  deleteMultipleFiles(
+    repoName: string,
+    items: Array<{ filePath: string; sha: string }>,
+    userId: SessionDataProps["isomerUserId"],
+    branchName: string
+  ): ResultAsync<string, GitFileSystemError | NotFoundError> {
+    let oldStateSha = ""
+    const efsVolPath = this.getEfsVolPathFromBranch(branchName)
+    const isStaging = this.isStagingFromBranchName(branchName)
+
+    return this.getLatestCommitOfBranch(repoName, branchName)
+      .andThen((latestCommit) => {
+        if (!latestCommit.sha) {
+          return errAsync(
+            new GitFileSystemError(
+              `Unable to find latest commit of repo: ${repoName} on branch "${branchName}"`
+            )
+          )
+        }
+        oldStateSha = latestCommit.sha as string
+        return okAsync(true)
+      })
+      .andThen(() =>
+        ResultAsync.combine(
+          items.map(({ filePath, sha }) =>
+            this.getFilePathStats(repoName, filePath, isStaging).andThen(
+              (stats) =>
+                okAsync({
+                  filePath,
+                  sha,
+                  stats,
+                })
+            )
+          )
+        )
+      )
+      .andThen((itemsWithStats) => {
+        const isFileConflict = !itemsWithStats.every(async (itemStats) => {
+          if (itemStats.stats.isDirectory()) {
+            // If it's a directory, skip the blob hash verification
+            return true
+          }
+
+          const result = await this.getGitBlobHash(
+            repoName,
+            itemStats.filePath,
+            isStaging
+          ).andThen((sha) => {
+            if (sha !== itemStats.sha) {
+              return okAsync(false)
+            }
+
+            return okAsync(true)
+          })
+
+          return result.isOk() && result.value
+        })
+
+        if (isFileConflict) {
+          return errAsync(
+            new ConflictError(
+              "File has been changed recently, please try again"
+            )
+          )
+        }
+
+        return okAsync(itemsWithStats)
+      })
+      .andThen((itemsWithStats) =>
+        // Note: All deletions must be successful, otherwise we rollback
```

Comment:
would this mean a single failure in a batch of 20 deletions will rollback all deletions? we don't support partial success of the request?

Reply:
This is more for simplicity. There can be many reasons why failure happens, so I decided to treat it as an atomic transaction instead of having some success some failure. On the frontend, we don't also have an easy way to show the user which files failed to delete, so I decided to go for getting the user to retry again.